content/post/walkthrough-exploiting-bugs-to-gain-code-execution.md

```diff
@@ -0,0 +1,338 @@
++++
+author = "cbrown"
+date = "2015-09-18T19:55:54-07:00"
+description = ""
+tags = ["exploit", "rop", "aslr", "shell"]
+title = "Walkthrough: Exploiting bugs to gain code execution on a modern operating system"
+
++++
+
+Whenever I discover a bug or typo in code I've written, I like to ask myself what the worst case effects of that typo might be. The most interesting question of course, is "If someone else found this before I did, could they have used it to gain arbitrary code execution?"
+
+We've seen arbitrary *root* code execution gained from as little as a [single random bit being erroneously flipped](https://www.blackhat.com/docs/us-15/materials/us-15-Seaborn-Exploiting-The-DRAM-Rowhammer-Bug-To-Gain-Kernel-Privileges-wp.pdf), so in cases of buffer overflow or use-after-free errors, I tend to think the answer is virtually always yes.
+
+But as an exercise, I occasionally feel a need to actually develop exploits, and if you're into C++, it's a great learning experience. Due to advances in binary security there are an increasing number of hurdles, but I'm gonna walk you through the steps to exploit a simple overflow on a modern operating system.
+
+**Prerequisites:** Basic knowledge of how a [call stack](https://en.wikipedia.org/wiki/Call_stack) works, basic understanding of common assembly instructions (push, pop, mov, add...). Being comfortable debugging executables is a plus.
```

Comment:
I don't think you need to change this, but I think using `and` instead of a comma might work better.